### PR TITLE
[WIP] Better support of Ruby 2.3 with enabled frozen string literal

### PR DESCRIPTION
**Pull request by deepj**
This commit prevents several bugs with `RuntimeError: can't modify frozen String` with enabled string frozen literal in Ruby 2.3

### CODE REVIEW

**Comment by segiddins** [comment]
This shouldn't be necessary as none of the bundler files have the frozen string literal magic comment.

**Comment by deepj** [comment]
Of course, but for example if I try run RSpec dev tests with enabled string frozen literal (I want to use RSpec with a project with Ruby 2.3+ and string frozen literal). Without these changes in Bundler I can't move further. These changes don't bring any backward incompatibilities. So why not?

**Comment by segiddins** [comment]
In that case, we need to test that bundler supports being run with frozen string literals enabled

**Comment by deepj** [comment]
OK. I take a look at Travis settings. And is it okay to disable `Style/EmptyLiteral` in Rubocop (https://travis-ci.org/bundler/bundler/jobs/95840769#L256)? Alternative is to use `''.dup`. But it seems `String.new` is preferred style. See: https://bugs.ruby-lang.org/issues/11759#note-12

**Comment by segiddins** [comment]
Yes, I definitely prefer `String.new`

**Comment by segiddins** [comment]
Hm, I can't even get the RubyGems specs to _run_ with frozen string literals enabled (.rubies/2.3.0-preview1/lib/ruby/2.3.0/x86_64-darwin15/rbconfig.rb:247:in `replace': can't modify frozen String (RuntimeError)). Given Ruby core itself isn't working... I don't know how much it's worth getting bundler to work with it enabled globally yet?

**Comment by deepj** [comment]
I forgot to mention, `ruby-head` needs some additional patches due to this. I have some fixes on my side. See: https://github.com/ruby/ruby/pulls (https://github.com/ruby/ruby/pull/1116)

Let's make this PR like WIP

**Comment by segiddins** [comment]
I think it also needs https://github.com/ruby/ruby/pull/1136

**Comment by segiddins** [comment]
`can't modify frozen String, created at /home/travis/.rvm/rubies/ruby-head/lib/ruby/site_ruby/2.3.0/rubygems/dependency.rb:309` :cry:

File: lib/bundler/lazy_specification.rb [hunk 1]
```diff
@@ -41,6 +41,8 @@ def to_lock
         out = "    #{name} (#{version}-#{platform})\n"
       end
 
+      out = out.dup
```
Comment (bquorning) [inline lib/bundler/lazy_specification.rb]: Could we set an `out = String.new` variable at the beginning of the method instead of calling `dup` here?

**Comment by bquorning** [comment]
Please, also fix these:

 - https://github.com/bundler/bundler/blob/master/lib/bundler/env.rb#L14
 - https://github.com/bundler/bundler/blob/master/lib/bundler/env.rb#L65

**Comment by mwpastore** [comment]
It's worth mentioning that Ruby 3 will have frozen string literals enabled by default—or at least that's the plan—so any effort put into making Bundler compatible with this pragma now will hopefully not go to waste. I for one am hoping to run my full stack under Ruby 2.3 with `RUBYOPT="--enable=frozen-string-literal"` once we're all past the initial shock of incompatibility.

**Comment by deepj** [comment]
@mwpastore Unfortunately, Ruby core/stdlib isn't prepared for that. Also it seems core developers aren't interesting in that to fix it :( /cc @hsbt

File: lib/bundler/lazy_specification.rb (hunk 1 above)
Comment (deepj) [inline lib/bundler/lazy_specification.rb]: Fixed

**Comment by deepj** [comment]
@bquorning This is fixed as well

**Comment by mwpastore** [comment]
@deepj Agreed, the state of the preview is quite dismal.

**Comment by deepj** [comment]
@mwpastore It seems they solved it by using adding `frozen_string_literal: false ` to everywhere. More here: https://github.com/ruby/ruby/commit/3e92b635fb5422207b7bbdc924e292e51e21f040

**Comment by deepj** [comment]
@segiddins: Shouldn't I add `frozen_string_literal: true` to everywhere in bundler and not force frozen string literals globally like now? This way is more viable than affecting other dependencies. You made same thing in rubygems (https://github.com/rubygems/rubygems/pull/1408)

**Comment by segiddins** [comment]
@deepj I really dislike encoding comments so id rather not, if possible

**Comment by gregd** [comment]
I've installed Ruby 2.3.0, set --enable-frozen-string-literal cmd line switch and run my Rails 4.2 app. It stops on `rubygems.rb:65:in block in to_lock: can't modify frozen String` It would be good have the pull request merged in next Bundler release.

**Comment by indirect** [comment]
@gregd if it's important to you, you're welcome to fix the tests on Ruby 2.3, and then update this pull request.

**Comment by technion** [comment]
Hi, Just wondering if there were any remaining issues for the current version of this PR? I'm working on running with frozen strings and, at first glance, every issue outstanding is already covered by this patch.

**Comment by gregd** [comment]
@technion Try to execute tests. This PR fixes just few places where frozen strings are modified.

**Comment by indirect** [comment]
@homu r+

**Pull request merged**